dot/services_test.go

```diff
@@ -0,0 +1,639 @@
+		net *network.Service
+	}
+	tests := []struct {
+		name string
+		args args
+		want *grandpa.Service
+		err  error
+	}{
+		{
+			name: "base case",
+			args: args{
+				cfg: cfg,
+				st:  stateSrvc,
+				ks:  ks.Gran,
+				net: networkSrvc,
+			},
+			want: &grandpa.Service{},
+			err:  nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			no := nodeBuilder{}
+			got, err := no.createGRANDPAService(tt.args.cfg, tt.args.st, nil, tt.args.ks, tt.args.net,
+				nil)
+			assert.ErrorIs(t, err, tt.err)
+			if tt.want != nil {
+				assert.NotNil(t, got)
+			} else {
+				assert.Nil(t, got)
+			}
+		})
+	}
+}
+
+func Test_createRuntime(t *testing.T) {
+	t.Parallel()
+	cfg := NewTestConfig(t)
+	stateSrvc := newStateService(t)
+	code, err := stateSrvc.Storage.LoadCode(nil)
+	require.NoError(t, err)
+
+	cfgLife := NewTestConfig(t)
+	cfgLife.Core.WasmInterpreter = life.Name
+
+	type args struct {
+		cfg  *Config
+		ns   runtime.NodeStorage
+		st   *state.Service
+		code []byte
+	}
+	tests := []struct {
+		name string
+		args args
+		want runtime.Instance
+		err  error
+	}{
+		{
+			name: "wasmer runtime",
+			args: args{
+				cfg:  cfg,
+				ns:   runtime.NodeStorage{},
+				st:   stateSrvc,
+				code: code,
+			},
+			want: &wasmer.Instance{},
+			err:  nil,
+		},
+		{
+			name: "wasmer life",
+			args: args{
+				cfg:  cfgLife,
+				ns:   runtime.NodeStorage{},
+				st:   stateSrvc,
+				code: code,
+			},
+			want: &life.Instance{},
+			err:  nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := createRuntime(tt.args.cfg, tt.args.ns, tt.args.st, nil, nil, tt.args.code)
+			assert.ErrorIs(t, err, tt.err)
+			if tt.want != nil {
+				assert.NotNil(t, got)
+			} else {
+				assert.Nil(t, got)
+			}
+		})
+	}
+}
+
+func Test_nodeBuilder_newSyncService(t *testing.T) {
+	t.Parallel()
+	cfg := NewTestConfig(t)
+	stateSrvc := newStateService(t)
+	finalityGadget := &grandpa.Service{}
+	type args struct {
+		cfg             *Config
+		st              *state.Service
+		fg              sync.FinalityGadget
+		verifier        *babe.VerificationManager
+		cs              *core.Service
+		net             *network.Service
+		telemetryMailer telemetry.Client
+	}
+	tests := []struct {
+		name string
+		args args
+		want *sync.Service
+		err  error
+	}{
+		{
+			name: "base case",
+			args: args{
+				cfg:             cfg,
+				st:              stateSrvc,
+				fg:              finalityGadget,
+				verifier:        nil,
+				cs:              nil,
+				net:             nil,
+				telemetryMailer: nil,
+			},
+			want: &sync.Service{},
+			err:  nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			no := nodeBuilder{}
+			got, err := no.newSyncService(tt.args.cfg, tt.args.st, tt.args.fg, tt.args.verifier, tt.args.cs,
+				tt.args.net, tt.args.telemetryMailer)
+			assert.ErrorIs(t, err, tt.err)
+			if tt.want != nil {
+				assert.NotNil(t, got)
```

Comment:
should be able to `assert.Equal` after changing `state.Service` into an interface.